crates/ruff_python_parser/src/string.rs

```diff
@@ -304,40 +305,27 @@ impl<'a> StringParser<'a> {
                     }
 
                     let ret = if self_documenting {
-                        // TODO: range is wrong but `self_documenting` needs revisiting beyond
-                        // ranges: https://github.com/astral-sh/ruff/issues/5970
-                        vec![
-                            Expr::from(ast::ExprConstant {
-                                value: Constant::Str(expression.clone() + "="),
-                                kind: None,
-                                range: self.range(start_location),
-                            }),
-                            Expr::from(ast::ExprConstant {
-                                value: trailing_seq.into(),
-                                kind: None,
-                                range: self.range(start_location),
-                            }),
-                            Expr::from(ast::ExprFormattedValue {
-                                value: Box::new(
-                                    parse_fstring_expr(&expression, start_location).map_err(
-                                        |e| {
-                                            FStringError::new(
-                                                InvalidExpression(Box::new(e.error)),
-                                                start_location,
-                                            )
-                                        },
-                                    )?,
-                                ),
-                                conversion: if conversion == ConversionFlag::None && spec.is_none()
-                                {
-                                    ConversionFlag::Repr
-                                } else {
-                                    conversion
-                                },
-                                format_spec: spec,
-                                range: self.range(start_location),
+                        let value =
+                            parse_fstring_expr(&expression, start_location).map_err(|e| {
+                                FStringError::new(
+                                    InvalidExpression(Box::new(e.error)),
+                                    start_location,
+                                )
+                            })?;
+                        let leading =
+                            &expression[..usize::from(value.range().start() - start_location) - 1];
+                        let before_eq =
+                            &expression[usize::from(value.range().end() - start_location) - 1..];
+                        vec![Expr::from(ast::ExprFormattedValue {
+                            value: Box::new(value),
+                            debug_text: Some(ast::DebugText {
+                                leading: leading.to_string(),
```

Comment:
Could we store references into the source here rather than allocating new strings? i guess many more things would need to change

Reply:
I played around with an ast that uses `Cow<'source, str>` for all string fields. I spent about 2h fixing lifetime issues and run out of patience. The problem is that every node must be parametrized by the source lifetime. 

We may still want to do this because using a bump allocator for the AST nodes would also require a new lifetime. Something to explore in the future.